File: scripts/algorithms/F/Flatten Nested List Iterator/Flatten Nested List Iterator.cpp
class NestedIterator {
    vector<int> v;
    int index;
    void helper(vector<NestedInteger> &nestedList, vector<int>& v)
    {
        for (int i = 0; i < nestedList.size(); i++)
        {
            if (nestedList[i].isInteger())
            {
                v.push_back(nestedList[i].getInteger());
            }
            else
            {
                helper(nestedList[i].getList(), v);
            }
        }
    }
public:
    NestedIterator(vector<NestedInteger> &nestedList) {
        helper(nestedList, v);
        index = 0;
    }
    
    int next() {
        return v[index++];
    }
    
    bool hasNext() {
        if (v.size() <= index)
        {
            return false;
        }
        return true;
    }
};

File: scripts/algorithms/F/Flatten Nested List Iterator/Flatten Nested List Iterator.java
// Runtime: 4 ms (Top 77.47%) | Memory: 46.1 MB (Top 81.31%)
public class NestedIterator implements Iterator<Integer> {
    List<Integer> list=new ArrayList();
    void flatten(List<NestedInteger> nestedList){
        for(NestedInteger nested:nestedList){
            if(nested.isInteger())
                list.add(nested.getInteger());
            else
                flatten(nested.getList());
        }
    }
    public NestedIterator(List<NestedInteger> nestedList) {
        flatten(nestedList);
    }
    int index=0;
    @Override
    public Integer next() {
        return list.get(index++);
    }

    @Override
    public boolean hasNext() {
        return index<list.size();
    }
}